sdk/azidentity/workload_identity.go

```diff
@@ -139,3 +156,197 @@ func (w *WorkloadIdentityCredential) getAssertion(context.Context) (string, erro
+		baseTransport:  baseTransport,
+	}
+	return nil
+}
+
+// customTokenEndpointTransport is a custom HTTP transport that redirects token requests
+// to the custom token endpoint when in custom token endpoint mode.
+//
+// All non-token requests are forwarded to the fallback client.
+type customTokenEndpointTransport struct {
+	caFile         string
+	caData         string
+	sniName        string
+	tokenEndpoint  *url.URL
+	fallbackClient *azcore.Client
+	baseTransport  *http.Transport
+}
+
+func (i *customTokenEndpointTransport) Do(req *http.Request) (*http.Response, error) {
+	if !i.isTokenRequest(req) {
+		// not a token request, fallback to the original transporter
+		return doForClient(i.fallbackClient, req)
+	}
+
+	tr, err := i.getTokenTransporter()
+	if err != nil {
+		return nil, err
+	}
+
+	newReq := req.Clone(req.Context())
+	newReq.URL.Scheme = i.tokenEndpoint.Scheme // this will always be https
+	newReq.URL.Host = i.tokenEndpoint.Host
+	newReq.URL.Path = i.tokenEndpoint.Path
+	newReq.Host = i.tokenEndpoint.Host
+
+	return tr.RoundTrip(newReq)
+}
+
+// loadCAPool loads the CA certificate pool from either file or data
+func (i *customTokenEndpointTransport) loadCAPool() (*x509.CertPool, error) {
+	var caDataBytes []byte
+	var err error
+
+	if i.caFile != "" {
+		caDataBytes, err = os.ReadFile(i.caFile)
+		if err != nil {
+			return nil, fmt.Errorf("read CA file %q: %w", i.caFile, err)
+		}
+	} else if i.caData != "" {
+		caDataBytes = []byte(i.caData)
+	} else {
+		return nil, fmt.Errorf("missing CA: neither CA file nor CA data provided")
+	}
+
+	// Error out if CA data is empty
+	if len(caDataBytes) == 0 {
+		if i.caFile != "" {
+			return nil, fmt.Errorf("CA file %q is empty", i.caFile)
+		} else {
+			return nil, fmt.Errorf("CA data is empty")
+		}
+	}
+
+	caPool := x509.NewCertPool()
+	if !caPool.AppendCertsFromPEM(caDataBytes) {
+		if i.caFile != "" {
+			return nil, fmt.Errorf("parse CA file %q: no valid certificates found", i.caFile)
+		} else {
+			return nil, fmt.Errorf("parse CA data: no valid certificates found")
+		}
+	}
+
+	return caPool, nil
+}
+
+// getTokenTransporter rebuilds the HTTP transport every time it's called.
+// This approach is acceptable because token requests are infrequent due to token caching at higher levels.
+func (i *customTokenEndpointTransport) getTokenTransporter() (*http.Transport, error) {
+	transport := i.baseTransport.Clone()
+
+	caPool, err := i.loadCAPool()
+	if err != nil {
+		return nil, err
+	}
+
+	if transport.TLSClientConfig == nil {
+		transport.TLSClientConfig = &tls.Config{}
+	}
+	transport.TLSClientConfig.RootCAs = caPool
+
+	// Configure SNI if specified
+	if i.sniName != "" {
+		transport.TLSClientConfig.ServerName = i.sniName
+	}
+
+	return transport, nil
+}
+
+const (
+	tokenRequestClientAssertionField     = "client_assertion"
+	tokenRequestClientAssertionTypeField = "client_assertion_type"
+)
+
+func (i *customTokenEndpointTransport) isTokenRequest(req *http.Request) bool {
```

Comment:
@enj 

in current MSAL implementation, it starts with a discovery call (`GET <tenant>/.well-known/openid-configuration`) for getting the token endpoint (cannot be disable), then perform token call on the token endpoint. Giving this sequence is outside of Azure Go SDK's control, I didn't go with the approach of modifying discovery call response (because we are not sure if it always happen). Instead, I updated the logic to check if both `client_assertion` and `client_assertion_type` present in the request body. This should be robust enough for detecting the token call from my understanding, WDYT?